File: enterprise-platform/enterprise-platform/svc-cart/src/http.js
// Thin HTTP client for service-to-service calls.
// Retries, timeout, and dependency metrics - the things that make
// inter-service calls survivable in production.
const axios = require('axios');
const logger = require('./logger');
const { dependencyRequestsTotal } = require('./metrics');

const TIMEOUT_MS = parseInt(process.env.HTTP_TIMEOUT_MS || '3000', 10);
const RETRIES = parseInt(process.env.HTTP_RETRIES || '2', 10);

async function callService(name, baseUrl, path, options = {}) {
  const url = `${baseUrl}${path}`;
  let lastErr;

  for (let attempt = 0; attempt <= RETRIES; attempt++) {
    try {
      const res = await axios({
        url,
        method: options.method || 'GET',
        data: options.data,
        headers: {
          'x-request-id': options.requestId || '',
          ...options.headers,
        },
        timeout: TIMEOUT_MS,
      });
      dependencyRequestsTotal.inc({ dependency: name, status: 'success' });
      return res.data;
    } catch (err) {
      lastErr = err;
      const status = err.response ? err.response.status : 'network_error';
      logger.warn({ dependency: name, url, attempt, status }, 'dependency call failed');

      // Don't retry client errors - they won't get better.
      if (err.response && err.response.status >= 400 && err.response.status < 500) {
        dependencyRequestsTotal.inc({ dependency: name, status: 'client_error' });
        throw err;
      }

      // Exponential backoff with jitter
      if (attempt < RETRIES) {
        const backoff = Math.min(100 * Math.pow(2, attempt), 1000);
        const jitter = Math.random() * 100;
        await new Promise((r) => setTimeout(r, backoff + jitter));
      }
    }
  }

  dependencyRequestsTotal.inc({ dependency: name, status: 'failure' });
  throw lastErr;
}

module.exports = { callService };

File: enterprise-platform/enterprise-platform/svc-payment/src/http.py
"""Service-to-service HTTP client with timeout, retry and metrics."""
import asyncio
import os
import random

import httpx

from .logger import log
from .metrics import DEPENDENCY_REQUESTS_TOTAL

TIMEOUT_S = float(os.getenv("HTTP_TIMEOUT_MS", "3000")) / 1000
RETRIES = int(os.getenv("HTTP_RETRIES", "2"))


async def call_service(name, base_url, path, method="GET", json=None, request_id=""):
    url = f"{base_url}{path}"
    last_err = None

    async with httpx.AsyncClient(timeout=TIMEOUT_S) as client:
        for attempt in range(RETRIES + 1):
            try:
                response = await client.request(
                    method, url, json=json, headers={"x-request-id": request_id}
                )
                response.raise_for_status()
                DEPENDENCY_REQUESTS_TOTAL.labels(dependency=name, status="success").inc()
                return response.json()
            except httpx.HTTPStatusError as err:
                last_err = err
                status = err.response.status_code
                log.warning(
                    "dependency call failed",
                    extra={"dependency": name, "url": url, "status": status},
                )
                # 4xx won't get better on retry.
                if 400 <= status < 500:
                    DEPENDENCY_REQUESTS_TOTAL.labels(
                        dependency=name, status="client_error"
                    ).inc()
                    raise
            except (httpx.RequestError, asyncio.TimeoutError) as err:
                last_err = err
                log.warning(
                    "dependency network error",
                    extra={"dependency": name, "url": url, "attempt": attempt},
                )

            if attempt < RETRIES:
                backoff = min(0.1 * (2**attempt), 1.0) + random.random() * 0.1
                await asyncio.sleep(backoff)

    DEPENDENCY_REQUESTS_TOTAL.labels(dependency=name, status="failure").inc()
    raise last_err
